### DISCUSSION TITLE
Encryption-Decryption interface for CRYSTALS-Kyber

### DISCUSSION THREAD

**Discussion by iteash** [open-quantum-safe/liboqs#1464] [Q&A]
Purpose: Find the Encryption/Decryption interface for the algorithm CRYSTALS-Kyber

Learning: Only Key Encapsulation/Decapsulation interface is provided by liboqs. Internally there is an indcca enc/dec function that is used but was not able to utilize that functionality properly.

Question: Is there an enc/dec interface provided by Liboqs. If not, does anyone know the rationale behind not providing this interface. In my learning it should be possible to do enc/dec operations using CRYSTALS-Kyber algo. Is my understanding incorrect? Is the enc/dec operation in pipeline and not yet taken up or there is some other fundamental reason for not having this interface.

**Comment by baentsch** [comment]
> If not, does anyone know the rationale behind not providing this interface. 

One goal for `liboqs` used to be to make available interfaces valid for/implementable by all PQ algorithms supported by `liboqs`.

Personally, I'd think it may be time to revisit this original goal as the set of algorithms supported by `liboqs` is shrinking and the "downstream" need for an encryption interface clearly exists. What do you think @dstebila @christianpaquin ?

Would you be willing to contribute an API proposal @iteash ? Ideally combined with a use (test!) case...

**Comment by dstebila** [comment]
Well there are several design choices to be made when converting a KEM into a PKE.  I'm not aware of anyone having written out a spec for doing so for Kyber.  Probably the right way to do this is plugging Kyber into [HPKE](https://datatracker.ietf.org/doc/html/rfc9180).  I'm not sure if the right place to do this is in liboqs, or at a higher level.  For example, does OpenSSL implement HPKE and can an OpenSSL provider register new HPKE algorithms?

**Reply by greengalaxy2016** [reply]
I've been thinking of plugging KEM algorithm in to JDK crypto package. that was why in another discussion thread, I asked if supporting OpenJDK is on the plan or not.

**Comment by baentsch** [comment]
> For example, does OpenSSL implement HPKE and can an OpenSSL provider register new HPKE algorithms?

Not that I knew (see [supported provider operations](https://www.openssl.org/docs/manmaster/man7/provider.html#Operations)). But we could surely work to change that if there were applications for this... Suggestions welcome, @iteash @igorbarshteyn

**Comment by iteash** [comment]
Hi Michael, Douglas,

First and foremost, sorry for the late reply.

Our team would love to build a proposal to plug Kyber into a PKE/HPKE
scheme if this has not been started yet. Dr. Amit Kumar Chauhan addressed
in the mail will be taking it forward from our side.

Thanks and regards,
Iteash

**Comment by emendir** [comment]
I would hugely welcome an interface for basic asymmetric encryption in liboqs.
Not higher-level constructs such as PKE or HPKE, but pure and simple asymmetric encryption.

It's great to have classes for the more complex hybrid encryption ready-to-go, but I find it quite a let-down that the basic, fundamental underlying asymmetric encryption functionality implementing the quantum-resistant algorithms, which is after all, what this library is about, is unavailable in the APIs.

I know most applications use hybrid encryption instead of pure asymmetric encryption, but there are plenty of use-cases for pure asymmetric encryption.

Furthermore, for educational purposes, I think it is vital that users are given access to the fundamental mechanisms underlying higher-level complex constructs such as KEMs.

Adding access to the basic quatum-resistant asymmetric encryption and decryption functions would also make the API more wholesome and complete.
I was quite surprised and disappointed when looking at this cryptography library to see that it has no encrypt and decrypt functions in its API!
